src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -222,51 +223,242 @@ private bool ScanDesignation(bool permitTuple)
+                    return this.CheckFeatureAvailability(node, MessageID.IDS_FeatureRecursivePatterns);
                 }
-                finally
+
+                if (parsePropertySubpattern())
                 {
-                    this.Release(ref resetPoint);
+                    parseDesignation();
+                    var node = _syntaxFactory.PropertyPattern(type, propertySubpattern, designation);
+                    return this.CheckFeatureAvailability(node, MessageID.IDS_FeatureRecursivePatterns);
+                }
+
+                if (type != null && parseDesignation())
+                {
+                    return _syntaxFactory.DeclarationPattern(type, designation);
                 }
+
+                this.Reset(ref resetPoint);
+                return this.ParseSubExpression(Precedence.Expression);
             }
-            else
+            finally
             {
-                // In places where a pattern is supported, we do not support tuple types
-                // due to both syntactic and semantic ambiguities between tuple types and positional patterns.
+                this.Release(ref resetPoint);
+            }
+        }
 
-                // But it still might be a pattern such as (operand is 3) or (operand is nameof(x))
-                node = this.ParseSubExpression(Precedence.Expression);
+        private PatternSyntax ParsePattern()
+        {
+            var node = ParseExpressionOrPattern(forCase: false);
+            switch (node)
+            {
+                case PatternSyntax pattern:
+                    return pattern;
+                case ExpressionSyntax expression:
+                    return _syntaxFactory.ConstantPattern(expression);
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(node);
             }
+        }
+
+        private PropertySubpatternSyntax ParsePropertySubpattern()
+        {
+            ParseSubpatternList(out var openBraceToken, out var subPatterns, out var closeBraceToken, SyntaxKind.OpenBraceToken, SyntaxKind.CloseBraceToken);
+            return _syntaxFactory.PropertySubpattern(openBraceToken, subPatterns, closeBraceToken);
+        }
+
+        private void ParseSubpatternList(
+            out SyntaxToken openToken,
+            out Microsoft.CodeAnalysis.Syntax.InternalSyntax.SeparatedSyntaxList<SubpatternElementSyntax> subPatterns,
+            out SyntaxToken closeToken,
+            SyntaxKind openKind,
+            SyntaxKind closeKind)
+        {
+            Debug.Assert(openKind == SyntaxKind.OpenParenToken || openKind == SyntaxKind.OpenBraceToken);
+            Debug.Assert(closeKind == SyntaxKind.CloseParenToken || closeKind == SyntaxKind.CloseBraceToken);
+            Debug.Assert((openKind == SyntaxKind.OpenParenToken) == (closeKind == SyntaxKind.CloseParenToken));
```

Comment:
>Debug.Assert((openKind == SyntaxKind.OpenParenToken) == (closeKind == SyntaxKind.CloseParenToken)); [](start = 12, length = 99)

Please add an assert that current token matches openKind.

Reply:
Recorded.

---
In reply to: [151815592](https://github.com/dotnet/roslyn/pull/23208#discussion_r151815592) [](ancestors = 151815592)